iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp

```diff
@@ -0,0 +1,113 @@
+///  *tiledx = c ? x : t_tiledx;
+///  *tiledy = c ? y : t_tiledy;
+/// }
+static void makeSwizzledId(Location loc, OpBuilder b, Value workgroupIdX,
+                           Value workgroupIdY, Value gridSizeX, Value gridSizeY,
+                           Value& SwizzledIdX, Value& SwizzledIdY,
+                           unsigned swizzleTile) {
+  Value zero = b.create<arith::ConstantIndexOp>(loc, 0);
+  Value tile = b.create<arith::ConstantIndexOp>(loc, swizzleTile);
+  Value yModTile = b.create<arith::RemUIOp>(loc, workgroupIdY, tile);
+  Value yDivTile = b.create<arith::DivUIOp>(loc, workgroupIdY, tile);
+  Value swizzleParam = b.create<arith::MulIOp>(loc, yModTile, gridSizeX);
+  Value swizzleParam2 =
+      b.create<arith::AddIOp>(loc, workgroupIdX, swizzleParam);
+  Value swizzleParam3 = b.create<arith::RemUIOp>(loc, swizzleParam2, tile);
+  Value swizzleParam4 = b.create<arith::MulIOp>(loc, yDivTile, tile);
+  Value unboundedSwizzledIdX =
+      b.create<arith::DivUIOp>(loc, swizzleParam2, tile);
+  Value unboundedSwizzledIdY =
+      b.create<arith::AddIOp>(loc, swizzleParam3, swizzleParam4);
+  Value gyModTile = b.create<arith::RemUIOp>(loc, gridSizeY, tile);
+  Value gyAddTile = b.create<arith::AddIOp>(loc, swizzleParam4, tile);
+  Value condition1 =
+      b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::ne, gyModTile, zero);
+  Value condition2 = b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::ugt,
+                                             gyAddTile, gridSizeY);
+  Value condition3 = b.create<arith::AndIOp>(loc, condition1, condition2);
+  SwizzledIdX = b.create<arith::SelectOp>(loc, condition3, workgroupIdX,
+                                          unboundedSwizzledIdX);
+  SwizzledIdY = b.create<arith::SelectOp>(loc, condition3, workgroupIdY,
+                                          unboundedSwizzledIdY);
+}
+namespace {
+struct WorkGroupSwizzlePass
+    : public WorkGroupSwizzleBase<WorkGroupSwizzlePass> {
+  WorkGroupSwizzlePass(unsigned swizzleLogTile)
+      : swizzleLogTile(swizzleLogTile) {}
+
+  void getDependentDialects(DialectRegistry& registry) const override {
+    registry.insert<AffineDialect>();
+  }
+  LogicalResult initializeOptions(StringRef options) override {
+    if (failed(Pass::initializeOptions(options))) {
+      return failure();
+    }
+    swizzleLogTile = logTile;
+    return success();
+  }
+  void runOnOperation() override {
+    if (swizzleLogTile == 0) return;
+    unsigned swizzleTile = pow(2, swizzleLogTile);
+    FuncOp funcOp = getOperation();
+    std::array<IREE::HAL::InterfaceWorkgroupIDOp, 2> oldWorkgroupIds;
+    bool xFound = false, yFound = false;
+    funcOp.walk([&](IREE::HAL::InterfaceWorkgroupIDOp idOp) {
+      unsigned index = idOp.dimension().getZExtValue();
+      if (index == 0) {
+        oldWorkgroupIds[index] = idOp;
+        xFound = true;
+      } else if (index == 1) {
+        oldWorkgroupIds[index] = idOp;
+        yFound = true;
+      }
```

Comment:
What is the expectation when `index == 2`?

Reply:
We leave index==2 alone if it exists and its not a problem if it doesnt.

Reply:
One thing we were careful with this swizzling logic is to not change the shape of the grid. That is why we do not have to worry about what's happening in the third dimension